[ML] Space aware trained models

jgowdyelastic opened the pull request:
Adds new a new ml saved object type `trained-model` to represent the trained models and enables them to be assigned to spaces.
When creating a new trained model saved object, if the model is related to an existing DFA job, the model is added to the same spaces as the job. Otherwise the current space is used.

### Misc changes

- Adds a scheduled task to run our saved object initialisation code every hour. This will create any missing saved objects for jobs or trained models.
This is needed because we do not know when trained models will be created from running DFA jobs and the model IDs cannot be predicted.

- Adds new capabilities to cover trained model actions. Previously we used DFA capabilities as they were closely related.

### Endpoints

- Adds new `ml/saved_objects/update_trained_models_spaces` endpoint for updating spaces for a trained model

- Adds new `ml/saved_objects/trained_models_spaces` endpoint for listing all trained models and their spaces, which the user is allowed to see.

- The `api/ml/saved_objects/status` endpoint now contains trained model status information.
This includes:
    - whether or not the real trained model exists (`trainedModelExists: boolean`)
    - whether or not a DFA jobs exists which is linked to the model (`dfaJobExists: boolean | null`, null if the model is not linked to a DFA job)
    - whether or not the saved object for the trained model exists ('savedObjectExits: boolean')
    - whether or not the DFA job has been reference in the saved object (`dfaJobReferenced: boolean | null`, null if the model is not linked to a DFA job)

- The `api/ml/saved_objects/sync` and `api/ml/saved_objects/initialize` endpoints now contain trained model saved object repair information.
Note, the structure of `initialize` has slightly altered to group saved objects by type. This fixes a ID duplication bug.

- Adds kibana endpoint to put trained models into es and updates the getter endpoints to allow no specification of model ID to return all trained models.
These are not specific to the spaces work, but were missing from the existing trained models endpoints.

### UI

- Adds a callout to the top of the trained model page, to inform the user of unsynced trained models:

![image](https://user-images.githubusercontent.com/22172091/154073364-4bdde5f6-7e62-43d2-ad0d-b92be9000067.png)


- Sync callout now shows saved objects in categories:

![image](https://user-images.githubusercontent.com/22172091/154089986-2e5357fc-f78b-429d-aac7-caa2ad7aeb50.png)


- Adds "Trained models" tab and list to the Stack Management ML page to allow the user to assign spaces to models:

![image](https://user-images.githubusercontent.com/22172091/154101902-4b1efbcc-2da3-4424-b424-81407cacbd5b.png)

![image](https://user-images.githubusercontent.com/22172091/154102626-0c819c2c-cb7b-4582-b392-df33deb92d5d.png)

### Follow up work

- [ ] Tests! API tests are needed for trained model creation, sync and initialisation.

- [x] Trained model deletion will need a similar spaces check to jobs. To stop users from deleting models from spaces they cannot see and to give them the choice to remove the model from the current space rather than really deleting it. - https://github.com/elastic/kibana/pull/127438

- [x] Renaming of the variable `jobType` which should now be something like `mlSavedObjectType`. This will produce a lot of churn and so was not included in this PR - https://github.com/elastic/kibana/pull/128059


-----


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))


- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

jgowdyelastic commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented on x-pack/plugins/ml/common/types/saved_objects.ts:
```diff
@@ -7,18 +7,31 @@
 }
 
+export type SyncResult = {
+  [jobType in JobType | TrainedModelType]?: {
```
`jobType` is used throughout the ML code base. It needs to change to something like `mlSavedObjectType` but that is a mouthful, so suggestions are welcome.

Also as changing it will be a simple find and replace task, I would like to do it in a specific followup PR to reduce churn.

peteharverson replied:
`mlSavedObjectType` sounds a good choice to me

jgowdyelastic commented on x-pack/plugins/ml/server/routes/trained_models.ts:
```diff
@@ -155,6 +184,44 @@ export function trainedModelsRoutes({ router, routeGuard }: RouteInitialization)
+      path: '/api/ml/trained_models/{modelId}',
+      validate: {
+        params: modelIdSchema,
+        body: schema.any(),
```
I really don't think we should be validating the shape of the trained models for endpoints which are just wrappers for es endpoints.
es will instantly complain if the model JSON is incorrect.

I know we do this validation on AD and DFA endpoints, but I think we should consider the idea of removing these validations too.
This is something to discuss outside of this PR.

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

jgowdyelastic commented on x-pack/plugins/ml/public/application/services/ml_api_service/trained_models.ts:
```diff
@@ -81,10 +78,10 @@ export function trainedModelsApiProvider(httpService: HttpService) {
      */
     getTrainedModelStats(modelId?: string | string[], params?: InferenceStatsQueryParams) {
-      const model = (Array.isArray(modelId) ? modelId.join(',') : modelId) || '_all';
+      const model = Array.isArray(modelId) ? modelId.join(',') : modelId;
```
I've changed this to no use `_all` when requesting all jobs to be consistent with all of our other client side endpoint functions.
using `_all` will work if calling kibana our endpoints, unless it is specified with other ids. in that case an error will be returned saying there are no jobs/models with the ID `_all`. This is inline with es behaviour.

afharo reviewed (approved):
Changes to `src/core/server/saved_objects/migrations/integration_tests/type_registrations.test.ts` LGTM!

jgowdyelastic commented:
@elasticmachine merge upstream

peteharverson commented on x-pack/plugins/ml/server/routes/saved_objects.ts:
```diff
@@ -197,6 +210,44 @@ export function savedObjectsRoutes(
+   *
+   * @api {post} /api/ml/saved_objects/update_trained_models_spaces Update what spaces trained models are assigned to
+   * @apiName UpdateTrainedModelsSpaces
+   * @apiDescription Update a list of trained models to add and/or remove them from given spaces
```
Need the closing `.` for the description.

peteharverson commented on x-pack/plugins/ml/server/routes/trained_models.ts:
```diff
@@ -155,6 +184,44 @@ export function trainedModelsRoutes({ router, routeGuard }: RouteInitialization)
+   *
+   * @api {put} /api/ml/trained_models/:modelId Put a trained model
+   * @apiName PutTrainedModel
+   * @apiDescription Adds a new trained model
```
Need the closing `.` for the description text.

peteharverson commented on x-pack/plugins/ml/server/saved_objects/initialization/initialization.ts:
```diff
@@ -41,7 +41,7 @@ export function jobSavedObjectsInitializationFactory(
         return;
       }
-
+      // return;
```
Is this comment needed?

peteharverson commented on x-pack/plugins/ml/server/saved_objects/service.ts:
```diff
@@ -181,6 +208,7 @@ export function jobSavedObjectServiceFactory(
     return jobObject;
   }
 
+  // CHANGE!!!!!!!!!! this looks like a dupe of _getJobObjects
```
Did you want to do this in this PR?

jgowdyelastic replied:
on second review the two functions are not the same. will remove the comment

peteharverson commented on x-pack/plugins/ml/server/saved_objects/service.ts:
```diff
@@ -366,6 +406,335 @@ export function jobSavedObjectServiceFactory(
+  async function _deleteTrainedModel(modelId: string) {
+    const [model] = await _getTrainedModelObjects(modelId);
+    if (model === undefined) {
+      throw new MLJobNotFound('job not found');
```
Do you need a new `MLModelNotFound` error type here, and the text needs to say `model not found` doesn't it?

darnautov commented on x-pack/plugins/ml/server/saved_objects/sync.ts:
```diff
@@ -75,22 +81,63 @@ export function syncSavedObjectsFactory(
+              if (mod === undefined) {
+                results.savedObjectsCreated[type]![modelId] = {
+                  success: false,
+
```
```suggestion
```

darnautov commented on x-pack/plugins/ml/server/saved_objects/sync.ts:
```diff
@@ -75,22 +81,63 @@ export function syncSavedObjectsFactory(
+        }
         if (simulate === true) {
-          results.savedObjectsCreated[job.jobId] = { success: true, type };
+          results.savedObjectsCreated[type]![job.jobId] = { success: true };
```
is it safe to use `!` here?

darnautov replied:
because `type` is a variable and TS can't statically analyse it

peteharverson commented:
I wonder if the text in the synchronization callout needs amending - the reference to 'jobs' is a bit confusing when looking at the trained models page, or synchronizing saved objects that are missing for models:

![image](https://user-images.githubusercontent.com/7405507/155108740-d08662de-4437-405c-ad5e-659d30268044.png)

Same applies to the references to 'jobs' in the flyout:

![image](https://user-images.githubusercontent.com/7405507/155108926-c5c17eb6-5fae-4d42-a63c-253949766fe3.png)

darnautov commented on x-pack/plugins/ml/server/saved_objects/sync_task.ts:
```diff
@@ -0,0 +1,141 @@
+  private log: { [l: string]: (t: string, e?: Error) => void };
+
+  constructor(logger: Logger) {
+    this.log = createLocalLogger(logger, `Task ${SAVED_OBJECTS_SYNC_TASK_ID}: `);
```
should we use the ML plugin logger? so if works with`plugins.ml` name config 
```
logging:
  loggers:
    - name: plugins.ml
      appenders: [default, console]
```

darnautov commented on x-pack/plugins/ml/server/saved_objects/sync_task.ts:
```diff
@@ -0,0 +1,141 @@
+  return {
+    info: (text: string) => logger.info(`${preText}${text}`),
+    warn: (text: string) => logger.warn(`${preText}${text}`),
+    error: (text: string, e: any) => logger.error(`${preText}${text}`, e),
```
```suggestion
    error: (text: string, e: Error) => logger.error(`${preText}${text}`, e),
```

jgowdyelastic commented on x-pack/plugins/ml/public/application/components/job_spaces_sync/sync_list.tsx:
```diff
@@ -159,21 +170,31 @@ const DatafeedsRemoved: FC<{ syncItems: SyncSavedObjectResponse }> = ({ syncItem
+}) => {
+  return (
+    <EuiAccordion id={id} buttonContent={title} paddingSize="l">
+      {Object.entries(results).map(([type, items]) => {
+        return (
+          <>
+            <EuiText size="s">
+              <div style={{ fontWeight: 'bold' }}>{type}</div>
+              <ul>
+                {Object.keys(items).map((item) => (
+                  <li key={item}>{item}</li>
+                ))}
+              </ul>
+            </EuiText>
+            <EuiSpacer size="s" />
+          </>
+        );
+      })}
```
funny this wasn't flagged as an error, they used to be

peteharverson commented on x-pack/plugins/ml/public/application/components/job_spaces_sync/job_spaces_sync_flyout.tsx:
```diff
@@ -153,13 +153,15 @@ export const JobSpacesSyncFlyout: FC<Props> = ({ onClose }) => {
-      } else if (error !== undefined) {
-        errorCount++;
-      }
+  Object.values(resp).forEach((result: SyncResult) => {
```
As discussed, after synchronization is complete, the Overview page doesn't refresh so the warning callout remains. Is it possible to force a refresh, as is happening on the Trained Models page.

![image](https://user-images.githubusercontent.com/7405507/155119615-2eeb7fa1-9caa-4edd-bc80-7ccb197507c2.png)

peteharverson commented on x-pack/plugins/ml/server/saved_objects/sync_task.ts:
```diff
@@ -0,0 +1,141 @@
+
+              this.log.info(
+                count
+                  ? `${count} ML saved objects synced`
```
Nit, can `objects` be output as `object` if `count === 1`, rather than `Task ML:saved-objects-sync-task: 1 ML saved objects synced`

jgowdyelastic commented:
@elasticmachine merge upstream

peteharverson commented on x-pack/plugins/ml/public/application/components/saved_objects_warning/saved_objects_warning.tsx:
```diff
@@ -93,7 +93,7 @@ export const SavedObjectsWarning: FC<Props> = ({ jobType, onCloseFlyout, forceRe
           <FormattedMessage
             id="xpack.ml.jobsList.missingSavedObjectWarning.description"
-            defaultMessage="Some jobs are missing or have incomplete saved objects. "
+            defaultMessage="Some jobs or trained models are missing or have incomplete saved objects. "
```
The text for the link should probably be changed too. `Synchronize your jobs and trained models` or `Sychronize objects`?

![image](https://user-images.githubusercontent.com/7405507/155368505-b74ebeea-4410-4d53-b93f-7b73b5ea24e0.png)

peteharverson reviewed (commented):
Tested after changes made in https://github.com/elastic/kibana/pull/123487/commits/ef0a53bb4a67d06d95dd76b5420def35ec04001f and other than the text edits suggested it LGTM. Will wait on @jportner reviewing before giving it a last look.

jportner commented:
> Tested after changes made in [ef0a53b](https://github.com/elastic/kibana/commit/ef0a53bb4a67d06d95dd76b5420def35ec04001f) and other than the text edits suggested it LGTM. Will wait on @jportner reviewing before giving it a last look.

Thanks, I'm wrapping up a bugfix for 8.1, planning to review this early next week

tylersmalley commented:
A heads up that the cloud deployment failed the migration with the following:

```
Unable to complete saved object migrations for the [.kibana] index. Please check the health of your Elasticsearch cluster and try again. Unexpected Elasticsearch ResponseError: statusCode: 400, method: PUT, url: /.kibana_8.2.0_001/_mapping?timeout=60s error: [illegal_argument_exception]: mapper [ml-trained-model.job.create_time] cannot be changed from type [long] to [date],
```

Not sure if it's due to a change in this PR or because the ES cluster was persisted and the type changed during development.

jportner reviewed (commented):
OK, I've looked it over from a Spaces/Security perspective, I have a few questions below.

> ### Follow up work
> ...
> * Trained model deletion will need a similar spaces check to jobs. To stop users from deleting models from spaces they cannot see and to give them the choice to remove the model from the current space rather than really deleting it.

(1) It seems like this should be included in the MVP because there's no way to disable Trained Models, and it doesn't sound like too much trouble to add the authorization checks in the `authorizationProvider`. Do you think it's feasible to address this in this PR? 

(2) I referenced the [ML docs](https://www.elastic.co/guide/en/machine-learning/8.0/setup.html#kib-security-privileges) and created a minimally-privileged Kibana role "ml-admin" to grant full access (All) to Machine Learning features.
When I use this role to navigate to the "Machine Learning > Model Management > Trained Models" page, I see no trained models, and I see this error in the Kibana console logs: 
```
[2022-02-28T14:38:23.922-05:00][ERROR][plugins.ml] ResponseError: security_exception: [security_exception] Reason: action [cluster:admin/ingest/pipeline/get] is unauthorized for user [test] with roles [ml-admin], this action is granted by the cluster privileges [read_pipeline,manage_ingest_pipelines,manage_pipeline,manage,all]
    at KibanaTransport.request (/Users/joe/GitHub/kibana-3/node_modules/@elastic/transport/src/Transport.ts:517:17)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

If this is expected, it seems like it should be either a legitimate authorization problem (we should show something in the UI), or an expected problem (we shouldn't log an error message).

Also, either way, we should probably update the docs accordingly.

(3) If I grant the above role the additional ES Cluster privilege `manage_ingest_pipelines`, I no longer see the Kibana error log, but I still don't see the trained model either. In fact, I don't see the trained model until I also give the role `All` privileges for Saved Objects Management. That tells me that one of the underlying Saved Objects Repository calls you are making (probably a `find`) relies on the user having privileges to access the new `'ml-trained-model'` saved object type. (Saved Object Management gives the user privileges to access _all_ saved object types).

----

I haven't figured out how to make _new_ trained models, I was just basing this off of the existing "lang_ident_model_1" trained model that already exists. But I'd be interested in getting a brief walkthrough of how to use trained models so I can try to help exercise other AuthZ corner cases.

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@jportner 

> (1) It seems like this should be included in the MVP because there's no way to disable Trained Models, and it doesn't sound like too much trouble to add the authorization checks in the `authorizationProvider`. Do you think it's feasible to address this in this PR?

The majority of this work will be in the UI where we control whether a job/model is deleted or untagged from a space.
Historically, from a kibana endpoint points of view, we allow a user to delete a job if they are able to see it. We only put restrictions on whether the job is in spaces they cannot see in the UI.
In the UI the user cannot delete a job if it exists in spaces the user cannot see or if it is in the `*` space and the user cannot see all spaces.

My intention was to add these UI checks and confirmation modals in a follow up PR that will make it in for 8.2 along with this PR. Mainly to reduce the size of this PR.
But if that is undesirable I can add the work in here.

> 
> (2) I referenced the [ML docs](https://www.elastic.co/guide/en/machine-learning/8.0/setup.html#kib-security-privileges) and created a minimally-privileged Kibana role "ml-admin" to grant full access (All) to Machine Learning features. When I use this role to navigate to the "Machine Learning > Model Management > Trained Models" page, I see no trained models, and I see this error in the Kibana console logs:

Yes I agree that this message shouldn't be logged like this. It is an acceptable situation, where the user doesn't have access to the ingest pipelines but they do have read access to the models.
I'll change it in this PR.

> (3) If I grant the above role the additional ES Cluster privilege `manage_ingest_pipelines`, I no longer see the Kibana error log, but I still don't see the trained model either. In fact, I don't see the trained model until I also give the role `All` privileges for Saved Objects Management. That tells me that one of the underlying Saved Objects Repository calls you are making (probably a `find`) relies on the user having privileges to access the new `'ml-trained-model'` saved object type. (Saved Object Management gives the user privileges to access _all_ saved object types).

This was a bug, I missed out adding the saved object type to the list of granted privileges for the ML plugin.
Added here  https://github.com/elastic/kibana/pull/123487/commits/e60baf5165bcba96939eb45bf0778429cb7d4363

jportner commented:
> My intention was to add these UI checks and confirmation modals in a follow up PR that will make it in for 8.2 along with this PR. Mainly to reduce the size of this PR.
> But if that is undesirable I can add the work in here.

That is generally my preference so we can review the PR in its entirety and make sure we aren't introducing authZ bugs, but if you're planning an immediate followup PR / before the 8.2 release, that seems reasonable!

> Yes I agree that this message shouldn't be logged like this. It is an acceptable situation, where the user doesn't have access to the ingest pipelines but they do have read access to the models. I'll change it in this PR. 

🎉 

> This was a bug, I missed out adding the saved object type to the list of granted privileges for the ML plugin.
> Added here [e60baf5](https://github.com/elastic/kibana/commit/e60baf5165bcba96939eb45bf0778429cb7d4363)

Great!

I also noticed an old TODO comment that can be removed 😅 I can't comment on it directly in this PR but it's here:

<img width="930" alt="image" src="https://user-images.githubusercontent.com/5295965/157092579-4f4b0f9b-8ee4-4e5c-9ac2-4d01742f460f.png">

----

Sounds like there's nothing else for me to review until your follow-up PR that adds the authZ checks! Let me know if you think otherwise.

jgowdyelastic commented:
> A heads up that the cloud deployment failed the migration with the following:
> 
> ```
> Unable to complete saved object migrations for the [.kibana] index. Please check the health of your Elasticsearch cluster and try again. Unexpected Elasticsearch ResponseError: statusCode: 400, method: PUT, url: /.kibana_8.2.0_001/_mapping?timeout=60s error: [illegal_argument_exception]: mapper [ml-trained-model.job.create_time] cannot be changed from type [long] to [date],
> ```
> 
> Not sure if it's due to a change in this PR or because the ES cluster was persisted and the type changed during development. I deleted the deployment and updated the branch to generate a new deployment to test that.

Thanks @tylersmalley , yes this was caused by a mid development saved object type change which must have caused a problem if the ES cluster has been persisted.

jgowdyelastic commented:
I will create a PR after this one is merged to add the space checks for deleting trained models.
I believe the existing jobs checks and UI components can be refactored to make them more generic for reuse.
Does this sound ok @peteharverson ?
cc @jportner

peteharverson reviewed (approved):
LGTM

Adding the space checks for deleting trained models in a follow-up PR sounds good to me. It will involve substantial code changes, so doing that work in a separate PR will make it easier to review. Same applies to the rename of the variable `jobType`.

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented on x-pack/plugins/ml/server/saved_objects/service.ts:
```diff
@@ -366,6 +406,335 @@ export function jobSavedObjectServiceFactory(
+  }
+
+  async function _createTrainedModel(modelId: string, job: TrainedModelJob | null) {
+    await isMlReady();
```
If i'm honest I don't think a change like that will add value or reduce code. Plus I believe the file would need to be refactored to be a class.
Decorators are nice for more complicated situations, this promise is just added to a handful of functions here to stop access to our saved objects before they have been initialised.

jgowdyelastic commented:
@elasticmachine merge upstream

jgowdyelastic commented:
@elasticmachine merge upstream

darnautov reviewed (approved):
LGTM

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 123487` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 123487` or prevent reminders by adding the `backport:skip` label.
